### DISCUSSION TITLE
Going previous screen on blazor hybrid apps

### DISCUSSION THREAD

**Discussion by mehmetilker** [dotnet/maui#7837] [Q&A]
There are questions about "going previous page/screen" when back button pressed on Android. It already works now, so handled by platform later releases I guess.

I was wondered if there is a solution for "going previous screen position"? When you slide little bit down, go to next page, and then when you go back, it goes top of the page.

Is there any solution about it?

**Comment by nextcodelab** [comment]
I use native Page for native navigation.
```xaml
<Grid>
        <b:BlazorWebView
            x:Name="rzor"
            BackgroundColor="{DynamicResource PageBackgroundColor}"
            HostPage="wwwroot/index.html"
            IsVisible="false">

            <b:BlazorWebView.RootComponents>
                <b:RootComponent ComponentType="{x:Type local:HighLightsPage}" Selector="#app" />
            </b:BlazorWebView.RootComponents>
        </b:BlazorWebView>
        <ActivityIndicator
            x:Name="activityIndicator"
            Background="transparent"
            BackgroundColor="Transparent"
            HeightRequest="60"
            HorizontalOptions="Center"
            IsRunning="True"
            IsVisible="true"
            VerticalOptions="Center"
            WidthRequest="60"
            Color="{DynamicResource PrimaryColor}" />
    </Grid>
```

**Comment by mehmetilker** [comment]
Hi @nextcodelab ,

So basically, what you are doing, instead of navigating in one BlazorWebView, you define new BlazorWebView for every page and navigate among them using NavigationManager?

**Reply by nextcodelab** [reply]
Yes this is a sample video
[https://user-images.githubusercontent.com/97447955/172508905-cb280db7-3b14-4e52-826e-80d76e235c34.mp4](https://user-images.githubusercontent.com/97447955/172508905-cb280db7-3b14-4e52-826e-80d76e235c34.mp4)

[https://github.com/dotnet/maui/discussions/7834](https://github.com/dotnet/maui/discussions/7834)

Even my shell flyout items is made up of html css..

**Comment by mehmetilker** [comment]
Can you say this way does not make any performance problems? Because there is an item in roadmap as stated in performance section:

-Enable pooling of WebViews and/or enable external creation of WebViews (from some externally-existing pool). This technique is used in major apps that often have multiple web views but only a subset of them are visible at one time.

https://visualstudiomagazine.com/articles/2022/02/10/blazor-hybrid-net-7.aspx

BTW, I guess that perf related item means Multiple WebView in same page. In my case it is multiple among View pages...

**Reply by nextcodelab** [reply]
I tried it my self, the collectionview is lag than webview... I used CollectionView to display wrap grid more 50 buttons and it loads 2 - 3 seconds,, but those 50 or more buttons is super lightning fast when using BlazorWebView almost click load instantly so I really need those webview than native controls for other purpose. But some of my Pages is real native like SettingsPage, AboutPage. 

Here the link of my app on the store almost every page is BlazorWebView except settings and about.

[https://play.google.com/store/apps/details?id=com.nxtccode.biblenotebook](https://play.google.com/store/apps/details?id=com.nxtccode.biblenotebook)

**Comment by mehmetilker** [comment]
So it won't be problem as long as there is not too much components in BlazorWebView.

Could you guide me how to implement navigation from one WebView .razor page to another WebView razor page?

I tried with the following way but does not work:

-I added new xaml file with following xml (apart from MainPage.xaml):

NewPag1.xaml
```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"             
             xmlns:local="clr-namespace:XApp.App"
             x:Class="XApp.App.NewPage1"
             Title="NewPage1">

    <BlazorWebView HostPage="wwwroot/index.html">
        <BlazorWebView.RootComponents>
            <RootComponent Selector="#app" ComponentType="{x:Type local:Main}" />
        </BlazorWebView.RootComponents>
    </BlazorWebView>
    
</ContentPage>
```

I am trying to navigate to that page like from index.razor:

```
    private void GoToAnotherPage()
    {
        App.Current.MainPage.Navigation.PushModalAsync(new NewPage1());
    }
```

It navigates to NewPage1 (I assume) but open index.razor again. How can I state to open, for example anotherpage.razor with parameters?

Thanks for your help.

**Reply by nextcodelab** [reply]
You can create single MainPrivacyPage.razor and consume it directly to xaml page.

```cs
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"             
             xmlns:local="clr-namespace:XApp.App"
             x:Class="XApp.App.NewPage1"
             Title="NewPage1">

    <BlazorWebView HostPage="wwwroot/index.html">
        <BlazorWebView.RootComponents>
            <RootComponent Selector="#app" ComponentType="{x:Type local:MainPrivacyPage.razor}" />
        </BlazorWebView.RootComponents>
    </BlazorWebView>
    
</ContentPage>
```
MainPrivacyPage.razor

```html
<div>
<H3> This is privacy page</H3>
</div>
@code{
}
```

**Reply by mehmetilker** [reply]
It did not work that way due to error: : 
XFC0000	Cannot resolve type "clr-namespace:XApp.App:About.razor"

But it worked when I added namespace: xmlns:pages="clr-namespace:Xappp.App.Pages"
and change local to pages like "<RootComponent Selector="#app" ComponentType="{x:Type pages:About}" />"

Problem with this approach, it does not load MainLayout... Loaded only About.razor content.

So I need to change the app as ShellApp and use xaml layout I guess.

Another question here is, how to pass parameter. For example loading Category.razor requires categorId parameters but how is this possible?

@nextcodelab sorry to keep you busy with my questions but it looks like there are lots of gaps to develop with hybrid approach....

**Reply by nextcodelab** [reply]
there are many methods to use like static object,, assign the object on static before navigate to that page, and get the file from static class.

example
```cs
public static SettingsXamlPage Current { get; private set; }

public object Parameter { get; }

public SettingsXamlPage(object parameter)
{
		InitializeComponent();
                Current = this;
		Parameter = parameter;
		
		
}
```
Consume on razor
SettingsRazorPage.razor

```cs
void OnInitialized()
{
      var parameter = SettingsXamlPage.Current.Parameter;
}
```

**Reply by mehmetilker** [reply]
Thanks @nextcodelab , it works. 
But I guess there should be more natural way to complete the flow. I will work on it.

There is a discussion here related with this problem.
https://github.com/dotnet/AspNetCore.Docs/issues/25872

So either there is a better solution and it is not documented yet, or no proper solution at all for now.
